rfcs/0027-frontier-synchronization.md

```diff
@@ -0,0 +1,188 @@
+             t
+          -> [ `Staged_ledger of Staged_ledger.t
+             | `Scan_state of Transaction_snark_scan_state.t ]
+      end
+    end
+  end
+
+  module Extensions : sig
+    type t
+
+    (* ... *)
+  end
+
+  module Read : sig
+    include Monad.S
+
+    (* to wrap computations in the monad *)
+    val deferred : 'a Deferred.t -> 'a t
+
+    val find : State_hash.t -> Breadcrumb.Any.Ref.t option t
+
+    (* necessary extension access will be provided here, Extensions.t will not be directly accessible here *)
+
+    (* ... *)
+  end
+
+  type t
+
+  (* Actions are representations of mutations (writes) to perform on
+   * the frontier. There is only one action that can be performed on
+   * a frontier right now, adding breadcrumbs, but we can add either
+   * a single breadcrumb, or a subtree of breadcrumbs. Future actions
+   * can also be added as needed *)
+  type action =
+    | Add_breadcrumb of Breadcrumb.Full.t
+    | Add_subtree of Breadcrumb.Full.t Rose_tree.t
+
+  val read :
+       t
+    -> f:('a Read.t)
+    -> 'a Deferred.t
+
+  val write :
+       t
+    -> f:(Action.t Read.t)
+    -> unit Deferred.t
+end
+```
+
+#### Usage Notes
+
+It's important that computations described in `'a Read.t` monad need to be short (in terms of execution time). Having long cycles ocurr when handling reads or writes will have a significant effect on the overall delay for updating the frontier. As such, code should be designed to return whatever values you want from the `'a Read.t` monad as early as possible, and to continue using those values outside of calls to `Transition_frontier.read`.
+
+## Drawbacks
+[drawbacks]: #drawbacks
+
+- current design does not address breadcrumb specific desync bugs fully
+- adds more complexity overhead to all frontier interactions
+- will have performance impact on the speed at which we update the frontier
+- may open up new vectors for adversarial DDoS attacks
+  - for instance, adversary selectively delays rebroadcasting blocks to target, then triggers multiple catchups (enqueuing many writes at once when the catchups jobs finish), and then flood the target read requests to make the writes as slow as possible (all this combined could take the target's stake offline for a period of time)
+- places new importance on reads being short (ideally synchronous) operations
+- read monad needs to be updated to expose new read functionalities for extensions
+  - preferrably, adding new extensions or functions to access extensions would not require changes to the read monad to expose them, but without higher kinded types, it's can't be done in a reasonable way
+    - e.g. w/ generalized polymorphic variants, extension ref type could be `type 'intf extension = ... constraint 'intf = _ [> ]` and then the function to access could be `val read_extension : ('result 'intf) extension -> 'result 'intf -> 'result t` and could be read like `let%bind x = read_extension Root_history (`Lookup y) in ...`
```

Comment:
you could probably do this with functors, though I think it's best to stay with what you're suggesting

Reply:
I don't think I understand how it could be implemented with functors. You can't have a signature in OCaml with a set of arbitrary members that you can extend. Each extension has their own set of functions for reading information from extensions, they don't have a unified interface for accessing values.